optuna/visualization/matplotlib/_optimization_history.py

```diff
@@ -128,11 +133,92 @@ def _get_optimization_history_plot(
         _logger.warning("Study instance does not contain trials.")
         return ax
 
-    ax = _get_optimization_histories(studies, target, target_name, ax)
+    if error_bar:
+        ax = _get_optimization_histories_with_error_bar(studies, target, target_name, ax)
+    else:
+        ax = _get_optimization_histories(studies, target, target_name, ax)
+
     plt.legend(bbox_to_anchor=(1.05, 1.0), loc="upper left")
     return ax
 
 
+def _get_optimization_histories_with_error_bar(
+    studies: List[Study],
+    target: Optional[Callable[[FrozenTrial], float]],
+    target_name: str,
+    ax: "Axes",
+) -> "Axes":
+
+    max_trial_number = np.max(
+        [
+            trial.number
+            for study in studies
+            for trial in study.get_trials(states=(TrialState.COMPLETE,))
+        ]
+    )
+
+    _target: Callable[[FrozenTrial], float]
+    if target is None:
+
+        def _target(t: FrozenTrial) -> float:
+            return cast(float, t.value)
+
+    else:
+        _target = target
+
+    target_values: List[List[float]] = [[] for _ in range(max_trial_number + 2)]
+    for study in studies:
+        trials = study.get_trials(states=(TrialState.COMPLETE,))
+        for t in trials:
+            target_values[t.number].append(_target(t))
+
+    mean_of_target_values = [np.mean(v) if len(v) > 0 else np.nan for v in target_values]
+    std_of_target_values = [np.std(v) if len(v) > 0 else np.nan for v in target_values]
+    trial_numbers = np.arange(max_trial_number + 2)[[v is not np.nan for v in mean_of_target_values]]
```

Comment:
could you take a look my final suggestion?
```suggestion
    trial_numbers = np.arange(max_trial_number + 2)[
        [v is not np.nan for v in mean_of_target_values]
    ]
```

Reply:
Indeed, it looks better.

Reply:
It is different from the plotly's implementation. If you prefer to use the same logic, please ignore my suggestion and apply @nzw0301's one (or we can change the original implementation as a follow-up).

Reply:
This is completely off-topic about this PR, but it is related to @himkt's comment, so let me leave my suggestion. 
When I review this PR, the logic to generate arguments of this plot function is really similar to ploty's one. So I'm wondering if we could merge the logic not depending on the backend of the packages: plotly/matplotlib.